File: insertion_merge/hybridsort.py
import time
import numpy as np
import math


def hybrid_mergesort(arr: list, start: int, end: int, thresh: int):
  # Sorts in place

    if start >= end:
        return

    if end-start+1 <= thresh:
        insertionsort(arr, start, end)
        return

    if end >= start + 2:
        mid = (start+end) // 2
        # print(f"mid is {mid}")
        hybrid_mergesort(arr, start, mid, thresh)
        hybrid_mergesort(arr, mid+1, end, thresh)

    merge(arr, start, end)


def merge(arr: list, start: int, end: int):
    if start >= end:
        return

    mid = start + (end-start) // 2
    start2 = mid + 1

    while start <= mid and start2 <= end:

        # Case 1: if first element is smaller than or equal to right element, i.e correct place
        if arr[start] <= arr[start2]:
            start += 1

        # Case 2: if first element is larger than right element, then shift the right element to before the left element
        else:
            # Right shift all elements from "start" up till element before "start2"
            value = arr[start2]
            for index in range(start2, start, -1):
                arr[index] = arr[index-1]
            # Copy over original start2 value into original start index
            arr[start] = value

            # Update all pointers for next iteration of while loop
            start += 1
            mid += 1
            start2 += 1


def insertionsort(arr: list, start: int, end: int):
    for i in range(start+1, end+1):
        for j in range(i, start, -1):
            if arr[j] < arr[j-1]:
                swap(arr, j, j-1)
            else:
                break


def swap(arr: list, i: int, j: int):
    tmp = arr[j]
    arr[j] = arr[i]
    arr[i] = tmp
    return


def time_hybridsort(arr: list, thresh: int):

    # start = time.process_time_ns()
    start = time.perf_counter_ns()
    hybrid_mergesort(arr, 0, len(arr)-1, thresh)
    # end = time.process_time_ns()
    end = time.perf_counter_ns()
    return end - start


def run_average_case():
    np.random.seed(27)

    for pow in range(2, 6):
        MIN_NUM = 0
        MAX_NUM = 999
        N = 10**pow
        ITERS = 1
        MAX_THRESH = 20

        matrix = np.zeros(MAX_THRESH)

        for i in range(ITERS):
            ori_arr = np.random.randint(low=MIN_NUM, high=MAX_NUM, size=N)
            for thresh in range(1, MAX_THRESH+1):
                arr = ori_arr.copy().tolist()
                # print(arr, end="")
                matrix[thresh-1] = time_hybridsort(arr, thresh)
                # print(f" --> {arr}")

        # Calculate average execution time across iterations for each threshold
        threshold_timings = np.mean(matrix, axis=0)

        # Save timings
        np.save(f"./timings/avr_pow{pow}.npy", threshold_timings)

        # print(threshold_timings)
        low = np.argmin(threshold_timings)
        print(f"N = {N}")
        print(
            f"Size=10^{pow}. Fastest at thresh={low+1} with speed {threshold_timings[low] / 1e9} seconds")


def run_best_case():
    for pow in range(2, 5):
        N = 10**pow
        ITERS = 3
        MAX_THRESH = 20

        matrix = np.zeros([ITERS, MAX_THRESH])

        for i in range(ITERS):
            ori_arr = [x for x in range(N)]
            for thresh in range(1, MAX_THRESH+1):
                arr = ori_arr[:]  # make a copy
                # print(arr, end="")
                matrix[i][thresh-1] = time_hybridsort(arr, thresh)
                # print(f" --> {arr}")

        # Calculate average execution time for each threshold
        threshold_timings = np.mean(matrix, axis=0)

        # Save timings
        np.save(f"./timings/best_pow{pow}.npy", threshold_timings)

        print(threshold_timings)
        low = np.argmin(threshold_timings)
        print(
            f"Size=10^{pow}. Fastest at thresh={low+1} with speed {threshold_timings[low] / 1e9} seconds")


def run_worst_case():
    for pow in range(2, 5):
        N = 10**pow
        ITERS = 3
        MAX_THRESH = 20

        matrix = np.zeros([ITERS, MAX_THRESH])

        for i in range(ITERS):
            ori_arr = worstCaseArrayOfSize(N)
            for thresh in range(1, MAX_THRESH+1):
                arr = ori_arr[:]  # make a copy
                # print(arr, end="")
                matrix[i][thresh-1] = time_hybridsort(arr, thresh)
                # print(f" --> {arr}")

        # Calculate average execution time for each threshold
        if ITERS > 1:
            threshold_timings = np.mean(matrix, axis=0)
        else:
            threshold_timings = matrix

        # Save timings
        np.save(f"./timings/worst_pow{pow}.npy", threshold_timings)

        print(threshold_timings)
        low = np.argmin(threshold_timings)
        print(
            f"Size=10^{pow}. Fastest at thresh={low+1} with speed {threshold_timings[low] / 1e9} seconds")


def worstCaseArrayOfSize(n):
    if n == 1:
        return [1]
    else:
        top = worstCaseArrayOfSize(int(math.floor(float(n) / 2)))
        bottom = worstCaseArrayOfSize(int(math.ceil(float(n) / 2)))
        return list(map(lambda x: x * 2, top)) + list(map(lambda x: x * 2 - 1, bottom))


if __name__ == "__main__":
    run_average_case()

File: insertion_merge/hybridsort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// #define THRESHOLD 20
#define N (int)1e5
#define MIN_NUM 0
#define MAX_NUM 999

void hybridsort(int *, int, int, int);
int compare(int, int);
void mergesort(int *, int, int);
void merge(int *, int, int);
void insertionsort(int *, int, int);
void swap(int *, int, int);
void printlist(int *, int);

int main()
{
  int best_threshold = -1;
  double best_cpu_time = 10000.00;

  clock_t start, end;
  double cpu_time_used;

  // int arr[] = {50, 10, 2, 70, 3, 5};
  // int arr[] = {9, 10, 11, 70, 3, 5};
  // int arr[] = {9, 8, 7, 6, 5, 4};
  // int arr[] = {4, 5, 6, 7, 8, 9};
  int original_arr[N];
  int variable_arr[N];
  srand(27);
  int range = MAX_NUM - MIN_NUM + 1;
  for (int i = 0; i < N; i++)
  {
    original_arr[i] = rand() % range + MIN_NUM; //assign random number to index in the array;
  }

  // printlist(arr, N);
  for (int threshold = 1; threshold < 50; threshold += 1)
  {
    // Reset the variable array
    for (int i = 0; i < N; i++)
    {
      variable_arr[i] = original_arr[i];
    }

    // Start the clock timer
    start = clock();
    hybridsort(variable_arr, 0, N - 1, threshold);
    end = clock();

    cpu_time_used = ((double)(end - start)) / CLOCKS_PER_SEC;
    if (cpu_time_used < best_cpu_time)
    {
      best_cpu_time = cpu_time_used;
      best_threshold = threshold;
    }
    // printlist(arr, N);
    printf("Threshold: %d, CPU time used: %.8f\n", threshold, cpu_time_used);
  }

  printf("Best threshold is %d with time of %.8f\n", best_threshold, best_cpu_time);

  return 0;
}

void printlist(int *arr, int length)
{
  for (int i = 0; i < length; i++)
  {
    printf("%d ", arr[i]);
  }
  printf("\n");
}

// start and end are indexes
void hybridsort(int *arr, int start, int end, int threshold)
{
  if (start >= end)
  {
    return;
  }

  // If below or equal to a certain threshold length, use Insertion Sort. Or else, use Merge Sort.
  int length = end - start + 1;
  if (length <= threshold)
  {
    insertionsort(arr, start, end);
    return; // do not proceed with mergesort code
  }

  if (start <= end - 2)
  {
    int mid = start + (end - start) / 2;
    // printf("mid is %d\n", mid);

    hybridsort(arr, start, mid, threshold);
    hybridsort(arr, mid + 1, end, threshold);
  }

  merge(arr, start, end);
}

void merge(int *arr, int start, int end)
{
  if (start >= end)
    return;

  // Merges arr from index n to index m
  int mid = start + (end - start) / 2;
  int a = start;
  int b = mid + 1;

  while (a <= mid && b <= end)
  {
    int cmp = compare(arr[a], arr[b]);

    // Case 1 : Right is greater than Left. Shift Right element to just before Left element.
    if (cmp > 0)
    {
      int tmp = arr[b]; // Save the value to be shifted
      for (int i = b; i > a; i--)
      {
        // Within the sorted portion, shift all values to the right. Leftmost value is unchanged.
        arr[i] = arr[i - 1];
      }
      arr[a] = tmp; // Save the tmp value into leftmost in sorted portion.

      // Increment all 3 pointers
      a++;
      b++;
      mid++;
    }
    else if (cmp < 0)
    {
      a++; // This value is in the right place. Just increment a by 1.
    }
    else
    {
      if (a == mid && b == end)
        break; // No movement needed if they are the last elements

      a++;              // Move a up by 1 because we want to insert the right side value only after this.
      int tmp = arr[b]; // Save the value to be shifted and move b forward by 1
      for (int i = b; i > a; i--)
      {
        // Within the sorted portion, shift all values to the right. Leftmost value is unchanged.
        arr[i] = arr[i - 1];
      }
      arr[a] = tmp;

      // Increment the 2 other pointers
      b++;
      mid++;
    }
  }
}

void insertionsort(int *arr, int start, int end)
{
  for (int i = start + 1; i <= end; i++)
  {
    // Check i-th element with left neighbour.

    // Case 1: i-th element is smaller than left neighbour
    if (arr[i] < arr[i - 1])
    {
      int counter = i;
      // Continuously swap the chosen element leftwards whenever possible. Stops when at correct position.
      while (counter > 0)
      {
        if (arr[counter] < arr[counter - 1])
        {
          swap(arr, counter, counter - 1);
        }
        else
        {
          break;
        }
        counter--;
      }
    }

    // Case 2: i-th element is greater than or equal to left neighobur
    // Nothing to be done
  }
}

void swap(int *arr, int i, int j)
{
  int tmp = arr[j];
  arr[j] = arr[i];
  arr[i] = tmp;
}

int compare(int val1, int val2)
{
  if (val1 > val2)
  {
    return 1;
  }
  else if (val1 == val2)
  {
    return 0;
  }
  else
  {
    return -1;
  }
}
